posthog/heatmaps/sql.py

```diff
@@ -0,0 +1,192 @@
+    --   * URL (maybe matching wild cards),
+    --   * width
+    -- we'll almost never query this by session id
+    -- so from least to most cardinality that's
+    ORDER BY (type, team_id,  toDate(timestamp), current_url, viewport_width)
+-- I am purposefully not setting index granularity
+-- the default is 8192, and we will be loading a lot of data
+-- per query, we tend to copy this 512 around the place but
+-- i don't think it applies here
+"""
+).format(
+    table_name=HEATMAPS_DATA_TABLE(),
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=HEATMAPS_DATA_TABLE_ENGINE(),
+)
+
+KAFKA_HEATMAPS_TABLE_SQL = lambda: KAFKA_HEATMAPS_TABLE_BASE_SQL.format(
+    table_name="kafka_heatmaps",
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=kafka_engine(topic=KAFKA_CLICKHOUSE_HEATMAP_EVENTS),
+)
+
+HEATMAPS_TABLE_MV_SQL = (
+    lambda: """
+CREATE MATERIALIZED VIEW IF NOT EXISTS heatmaps_mv ON CLUSTER '{cluster}'
+TO {database}.{target_table}
+AS SELECT
+    session_id,
+    team_id,
+    distinct_id,
+    timestamp,
+    -- x is the x with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    x,
+    -- y is the y with resolution applied, the resolution converts high fidelity mouse positions into an NxN grid
+    y,
+    -- stored so that in future we can support other resolutions
+    scale_factor,
+    viewport_width,
+    viewport_height,
+    -- some elements move when the page scrolls, others do not
+    pointer_target_fixed,
+    current_url,
+    type,
+    _timestamp
+FROM {database}.kafka_heatmaps
+""".format(
+        target_table="writable_heatmaps",
+        cluster=settings.CLICKHOUSE_CLUSTER,
+        database=settings.CLICKHOUSE_DATABASE,
+    )
+)
+
+# Distributed engine tables are only created if CLICKHOUSE_REPLICATED
+
+# This table is responsible for writing to sharded_heatmaps based on a sharding key.
+WRITABLE_HEATMAPS_TABLE_SQL = lambda: HEATMAPS_TABLE_BASE_SQL.format(
+    table_name="writable_heatmaps",
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=Distributed(
+        data_table=HEATMAPS_DATA_TABLE(),
+        sharding_key="cityHash64(concat(toString(team_id), '-', session_id, '-', toString(toDate(timestamp))))",
+    ),
+)
+
+# This table is responsible for reading from heatmaps on a cluster setting
+DISTRIBUTED_HEATMAPS_TABLE_SQL = lambda: HEATMAPS_TABLE_BASE_SQL.format(
+    table_name="heatmaps",
+    cluster=settings.CLICKHOUSE_CLUSTER,
+    engine=Distributed(
+        data_table=HEATMAPS_DATA_TABLE(),
+        sharding_key="cityHash64(concat(toString(team_id), '-', session_id, '-', toString(toDate(timestamp))))",
+    ),
+)
+
+DROP_HEATMAPS_TABLE_SQL = lambda: (
+    f"DROP TABLE IF EXISTS {HEATMAPS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
+)
+
+TRUNCATE_HEATMAPS_TABLE_SQL = lambda: (
+    f"TRUNCATE TABLE IF EXISTS {HEATMAPS_DATA_TABLE()} ON CLUSTER '{settings.CLICKHOUSE_CLUSTER}'"
+)
+
+INSERT_SINGLE_HEATMAP_EVENT = """
```

Comment:
This is a scary bit of text - as long as it's only ever used for testing we are good!